### PR TITLE
Wrap PublicKey GenericArray and add formatter

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#794

### CODE REVIEW

File: src/signature.rs [hunk 1]
```diff
@@ -11,11 +12,26 @@ use ring::{rand, signature};
...
 use std::cell::RefCell;
 use std::error;
+use std::fmt;
 use std::fs::File;
 use untrusted::Input;
 
 pub type KeyPair = Ed25519KeyPair;
-pub type PublicKey = GenericArray<u8, U32>;
+#[derive(Serialize, Deserialize, Clone, Copy, Default, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct PublicKey(pub GenericArray<u8, U32>);
```
Comment (garious) [inline src/signature.rs]: That GenericArray shouldn't be `pub`.

File: src/transaction.rs [hunk 2]
```diff
@@ -192,7 +192,7 @@ impl Transaction {
     /// Verify only the transaction signature.
     pub fn verify_sig(&self) -> bool {
         warn!("transaction signature verification called");
-        self.sig.verify(&self.from, &self.get_sign_data())
+        self.sig.verify(&self.from.0, &self.get_sign_data())
```
Comment (garious) [inline src/transaction.rs]: Instead, implement `AsRef`

Comment (garious) [reply src/transaction.rs]: Something like this, I'd think:

```rust
impl<T> AsRef<[T]> for PublicKey<T>
{
    fn as_ref(&self) -> &[T] {
        &self.0[..]
    }
}
```

Comment (CriesofCarrots) [reply src/transaction.rs]: Oh, nice!

**Comment by garious** [comment]
@aeyakovenko, @rob-solana, this look good to you guys? If so, we can do the same for Signature and Hash. And then we can circle back, and remove the Copy trait from each, tweak a bunch of function signatures, and possibly see improved performance by not copying all those long arrays throughout the codebase.

**Comment by garious** [comment]
@CriesofCarrots, Needs a rebase

**Comment by garious** [comment]
One more rebase please - to get CI to pass.

**Comment by CriesofCarrots** [comment]
@garious , done.
stable-perf check seems to be stuck on "scheduled"...?

File: src/signature.rs [hunk 3]
```diff
@@ -11,11 +12,38 @@ use ring::{rand, signature};
...
+#[derive(Serialize, Deserialize, Clone, Copy, Default, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct PublicKey(GenericArray<u8, U32>);
+
+impl PublicKey {
+    pub fn new(pubkey_vec: &[u8]) -> Self {
+        PublicKey(GenericArray::clone_from_slice(&pubkey_vec))
+    }
+}
+
+impl AsRef<GenericArray<u8, U32>> for PublicKey {
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply src/signature.rs]: Fair enough. Just edit the AsRef for now? Or start the migration to fixed-sized arrays?

[the remark this answers is not in the document]
Comment (garious) [reply src/signature.rs]: Right, just the AsRef. I just want to control the public interface to that data right now.

**Comment by CriesofCarrots** [comment]
@garious , are you okay with this AsRef now? CI failure should be resolved by Vines' last commit (1c8602e on #824)
I have Signature and Hash ready to go on this model.

**Comment by mvines** [comment]
landed #824, so hopefully if you re-trigger this PR it'll turn nice and green :)

**Comment by garious** [comment]
rebase, not retrigger, right?

**Comment by CriesofCarrots** [comment]
Is there a way to retrigger without committing new code in the pr? I've been wondering that.

**Comment by mvines** [comment]
close/open the PR is what I like to do

File: src/bin/wallet.rs [hunk 4]
```diff
@@ -3,6 +3,7 @@ extern crate bincode;
 extern crate bs58;
 extern crate clap;
 extern crate dirs;
+extern crate generic_array;
```
Comment (garious) [inline src/bin/wallet.rs]: Can that be deleted?

File: src/bin/wallet.rs [hunk 5]
```diff
@@ -3,6 +3,7 @@ extern crate bincode;
 extern crate bs58;
```
Comment (garious) [inline src/bin/wallet.rs]: Adding this to lib makes any of the ones in `bin/` redundant. Can you delete them?

**Pull request merged**